src/tools/cephfs/top/cephfs-top

```diff
@@ -0,0 +1,294 @@
+#!/usr/bin/python3
+
+import argparse
+import sys
+import curses
+import errno
+import json
+import signal
+import time
+
+from collections import OrderedDict
+from datetime import datetime
+from enum import Enum, unique
+
+import rados
+
+
+class FSTopException(Exception):
+    def __init__(self, msg=''):
+        self.error_msg = msg
+
+    def get_error_msg(self):
+        return self.error_msg
+
+
+@unique
+class MetricType(Enum):
+    METRIC_TYPE_NONE = 0
+    METRIC_TYPE_PERCENTAGE = 1
+    METRIC_TYPE_LATENCY = 2
+
+
+FS_TOP_PROG_STR = 'fstop'
+FS_TOP_PROG_VER = '0.0.1'
+
+# version match b/w fstop and stats emitted by mgr/stats
+FS_TOP_SUPPORTED_VER = 1
+
+ITEMS_PAD_LEN = 1
+ITEMS_PAD = "  " * ITEMS_PAD_LEN
+
+# metadata provided by mgr/stats
+FS_TOP_MAIN_WINDOW_COL_CLIENT_ID = "CLIENT_ID"
+FS_TOP_MAIN_WINDOW_COL_MNTPT = "MOUNT_POINT"
+FS_TOP_MAIN_WINDOW_COL_MNT_ROOT = "MOUNT_ROOT"
+FS_TOP_MAIN_WINDOW_COL_ADDR_HOST = "ADDR/HOST"
+
+MAIN_WINDOW_TOP_LINE_ITEMS_START = [ITEMS_PAD,
+                                    FS_TOP_MAIN_WINDOW_COL_CLIENT_ID,
+                                    FS_TOP_MAIN_WINDOW_COL_MNTPT,
+                                    FS_TOP_MAIN_WINDOW_COL_MNT_ROOT]
+MAIN_WINDOW_TOP_LINE_ITEMS_END = [FS_TOP_MAIN_WINDOW_COL_ADDR_HOST]
+
+# adjust this map according to stats version and maintain order
+# as emitted by mgr/stast
+MAIN_WINDOW_TOP_LINE_METRICS = OrderedDict([
+    ("CAP_HIT", MetricType.METRIC_TYPE_PERCENTAGE),
+    ("READ_LATENCY", MetricType.METRIC_TYPE_LATENCY),
+    ("WRITE_LATENCY", MetricType.METRIC_TYPE_LATENCY),
```

Comment:
This is the cumulative value, right? Do we have the average?

Reply:
Right. I don't think we collect avg stats at the moment.

@lxbsz Does the kclient send avg r/w/md latencies?

Reply:
(even if the kclient does, we do not collect it on the MDS atm)

Reply:
> Right. I don't think we collect avg stats at the moment.
>

Yeah, right, not for kclient as we disscussed.
 
> @lxbsz Does the kclient send avg r/w/md latencies?

Kclient for now will only sent the sum latency for all r/w/md. The same with libcephfs.

Reply:
ACK. I'll create a tracker for tracking this. (the sum latency is not much informative).

Reply:
Okay, I will update the kclient as needed then.

Reply:
https://tracker.ceph.com/issues/48619

I'll take care of user space bits.

Reply:
Copied it for kclient: https://tracker.ceph.com/issues/48621

As I remembered, why the client/kclient are sending the sum latency for them is that the mds side could collect the total OP numbers, doesn't work now ?

Reply:
(not sure if I understand it) do you mean the MDS just adding up and sum latencies?

IIRC, the client already used to track cumulative latencies and we started with those set of metrics (at least in user-space part).

Reply:
The total numbers of the r/w/md OPs has been recorded by MDS (?) And then the average latencies could be computed in the MDS side or by the cephfs-top tool with the sum latencies from client side.

That's what I was told about this before as I remembered(but not very sure now), the first patch series in kclient I had tried to send the average latencies and later I changed them to sum latencies.

Not very sure.

Reply:
> The total numbers of the r/w/md OPs has been recorded by MDS (?) And then the average latencies could be computed in the MDS side or by the cephfs-top tool with the sum latencies from client side.

Wouldn't using sum latencies to calculate avg latency be somewhat incorrect (or rather skewed)?

> 
> That's what I was told about this before as I remembered(but not very sure now), the first patch series in kclient I had tried to send the average latencies and later I changed them to sum latencies.
> 
> Not very sure.

Yeh -- I think we went back-and-forth on this in that PR. So the kclient just followed what the userspace client did.

Reply:
Okay, BTW, should we report the max/min latencies and also the standard deviation ?

Reply:
> Okay, BTW, should we report the max/min latencies and also the standard deviation ? Will it make sense ?

makes sense. I'll update the tracker so that I do not forget...

Reply:
Okay, BTW, the kclient have already support it.

Reply:
And also please help review https://github.com/ceph/ceph/pull/37945, which is new for opened files/inodes and pin inodes metrics when you are fee. Thanks in advance.